src/assertions/threshold.jl

```diff
@@ -0,0 +1,38 @@
+"""
+    Threshold(ratio; weights=nothing)
+
+Assert that the ratio of missing values in the provided dataset does not exceed to specified ratio.
+If a weights array is provided then the ratio will be calculated as the
+`sum(weights[ismissing.(data)]) / sum(weights)`
+"""
+struct Threshold <: Assertion
+    ratio::Float64
+    weights::Union{AbstractVector, Nothing}
+end
+
+Threshold(ratio::Float64; weights=nothing) = Threshold(ratio, weights)
+
+function assert(data::AbstractVector{Union{T, Missing}}, t::Threshold) where T
+    if t.weights === nothing
+        mratio = count(ismissing, data) / length(data)
+        if mratio > t.ratio
+            throw(AssertionError("Ratio of missing values exceeded $(t.ratio) ($mratio)."))
```

Comment:
I am going to get very confused to see this since:
```julia
julia> @assert 1==2
ERROR: AssertionError: 1 == 2
```

We using `MissingException` for this in a lot of code.
Could we use that here also?

Ideally the exception would have `ratio` and `mratio` as fields.
So that they could be programatically inspected.
That suggests the introduction of a `ThresholdException` that does this as part of its `Base.showerr` overload.

Reply:
We used to have `ImputeError` in the `Context` code, but since we aren't doing imputation here that seemed a bit misleading. using a `ThresholdException` would also cleanup this code a bit too 👍